crypto/evp/evp_locl.h

```diff
@@ -95,3 +95,67 @@ void *evp_generic_fetch(OPENSSL_CTX *ctx, int operation_id,
                                             OSSL_PROVIDER *prov),
                         int (*up_ref_method)(void *),
                         void (*free_method)(void *));
+
+/* Helper macros to avoid duplicating code */
+#define getparams(_ciph,_ptr,_sz,_key,_datatype,_statvar,_fn,_args)     \
```

Comment:
Make this a function. By making it a macro you *are* duplicating code. :)  And then the function can return a status directly, not set a ```_statvar``` and many of the little functions can be ```return getparams...``` etc.

Reply:
Then I'll have to reduce it to just param building, which has its own caveats, and have the get_params / set_params calls made explicitly in each spot this macro is currently used.  Half of its raison d'être goes [poof]

Reply:
> better still, just inline the code explicitly

Explicit code copying, that thing that leads to having to do the same bug fix later on on all copies, where one might miss one or two of them.  Right.  I'm not a fan.  (I actually had a bug in one macro, and I was glad I only had to fix it in one spot).

For the rest, ok I heard you, will do.

Reply:
Side note: oh how I lament that you can't pass down anonymous functions with closures in C.  That would make this kind of thing *so* much easier, and cleaner.

Reply:
Re this:

``` C
        return F(params);
```

You're missing the fact that some `F` take a `ctx` argument as well, while others don't.

Reply:
My point is that there isn't a lot of code to copy: check for NULL pointers, build a one param array and call a function.

Changing the duplicated code is just a `:%s/xxx/yyy/` in vi and in the TECO front end: `ESC ctrl-V META-x global-change ESC X META-c everywhere-but-these-places META-x something-or-other ESC more-arguments [lots of closing parentheses] ctrl-return` 😜

To avoid confusion, I can change it to `return F(...);.  With these inlined, each can be correct for its usage.  As utility functions, a bit more is required.

And yes, closures would be a huge benefit.  I like closures.

Reply:
TECO...  you had to mention TECO...